### ISSUE TITLE
自定义的sgd函数中的“除以batch_size”是否要省略需要视loss如何定义

### ISSUE THREAD

**Issue by Tianhonghai** [ShusenTang/Dive-into-DL-PyTorch#75]
**bug描述**
```
def sgd(params, lr, batch_size):
    # 为了和原书保持一致，这里除以了batch_size，但是应该是不用除的，因为一般用PyTorch计算loss时就默认已经沿batch维求了平均了。除batch size后训练速度变慢
    for param in params:
        param.data -= lr * param.grad #/ batch_size
```
首先说batch_size需要除就必须除，不能除就必须不除，没有应该不应该。

loss的定义如果是如3.7节中调用nn.CrossEntropyLoss()，是**不能**除以batch_size的，原因如源码所述，CrossEntropyLoss()已经沿batch_size取了平均。
而如果loss是采用3.6节中自定义的cross_entropy函数，而且在训练时l = loss(y_hat, y).sum()（计算的是批量样本loss和，不是平均），那在sgd中**必须**要除以batch_size，而如果定义l = loss(y_hat,     y).sum()/batch_size，效果跟调用nn.CrossEntropyLoss()是相同的，此时sgd就**不能**再除以batch_size了。

这里定义loss的方式不同，带来的另一个问题是每一个epoch结束后计算平均loss的方式，如#74 所述，如果loss的定义是3.7节中调用nn.CrossEntropyLoss()，在print loss时，应该是train_loss_sum/len(train_iter)，而如果l = loss(y_hat, y).sum()，才是train_loss_sum/n

**版本信息**
pytorch:1.3.1
torchvision:0.4.2
torchtext:0.4.0
...

**Comment by ShusenTang** [comment]
道理都懂，所以我注明了”为了和原书保持一致，这里除以了batch_size，但是应该是不用除的，因为一般用PyTorch计算loss时就默认已经沿batch维求了平均了“，另外在3.9.6节也说明了这个问题。但是由于改的话要改的章节太多，所以目前不打算改，等更新完毕后再整体改。

**Comment by Tianhonghai** [comment]
TKS
sgd里如果把/batch_size删掉的话，会有梯度爆炸，出不来结果的。

**Comment by ShusenTang** [comment]
> 出不来结果的

是不是你算loss的时候没有除以batchsize。反正要除一次batchsize，要么是在算loss的时候除，要么是在sgd里面除
